sdk/src/client/api/block_builder/input_selection/requirement/amount.rs

```diff
@@ -72,379 +121,136 @@ impl InputSelection {
+        let (inputs_sum, mut outputs_sum, _, _) = self.amount_sums();
+        let (remainder_amount, native_tokens_remainder, mana_remainder) = self.required_remainder_amount()?;
+        if inputs_sum > outputs_sum {
+            let diff = inputs_sum - outputs_sum;
 
-            if self.remainder_amount > diff {
-                self.remainder_amount - diff
-            } else {
-                0
+            if remainder_amount > diff {
+                outputs_sum += remainder_amount - diff
             }
-        } else if self.inputs_sum < self.outputs_sum {
-            self.outputs_sum - self.inputs_sum
-        } else if self.native_tokens_remainder || self.mana_remainder {
-            self.remainder_amount
-        } else {
-            0
+        } else if native_tokens_remainder || mana_remainder {
+            outputs_sum += remainder_amount
         }
+        Ok((inputs_sum, outputs_sum))
     }
 
-    fn fulfil<'a>(
-        &mut self,
-        input_selection: &InputSelection,
-        inputs: impl Iterator<Item = &'a InputSigningData>,
-    ) -> Result<bool, Error> {
-        for input in inputs {
-            if self.newly_selected_inputs.contains_key(input.output_id()) {
-                continue;
-            }
-
-            if let Some(sdruc) =
-                sdruc_not_expired(&input.output, input_selection.latest_slot_commitment_id.slot_index())
-            {
-                // Skip if no additional amount is made available
-                if input.output.amount() == sdruc.amount() {
-                    continue;
-                }
-                let input_sdr = self.inputs_sdr.get(sdruc.return_address()).unwrap_or(&0) + sdruc.amount();
-                let output_sdr = *self.outputs_sdr.get(sdruc.return_address()).unwrap_or(&0);
-
-                if input_sdr > output_sdr {
-                    let diff = input_sdr - output_sdr;
-                    self.outputs_sum += diff;
-                    *self.outputs_sdr.entry(sdruc.return_address().clone()).or_default() += sdruc.amount();
-                }
-
-                *self.inputs_sdr.entry(sdruc.return_address().clone()).or_default() += sdruc.amount();
-            }
-
-            if let Some(nt) = input.output.native_token() {
-                self.selected_native_tokens.insert(*nt.token_id());
-            }
-
-            self.inputs_sum += input.output.amount();
-            self.newly_selected_inputs.insert(*input.output_id(), input.clone());
-
-            if input.output.native_token().is_some() {
-                // Recalculate the remaining amount, as a new native token may require a new remainder output.
-                let (remainder_amount, native_tokens_remainder, mana_remainder) =
-                    self.remainder_amount(input_selection)?;
+    fn reduce_funds_of_chains(&mut self, input_amount: u64, output_amount: &mut u64) -> Result<bool, Error> {
+        if *output_amount > input_amount {
```

Comment:
This looks a bit weird, shouldn't this be used to decide if this function here should even be called?

Reply:
I debated this a lot, but it was slightly cleaner this way